guvnor-ala/guvnor-ala-services-rest/src/main/java/org/guvnor/ala/services/rest/RestRuntimeProvisioningServiceImpl.java

```diff
@@ -28,118 +32,365 @@
+            return new RuntimeQueryResultItemBuilder();
+        }
+
+        public RuntimeQueryResultItemBuilder with(PipelineExecutorTrace pipelineExecutorTrace) {
+            this.pipelineExecutorTrace = pipelineExecutorTrace;
+            return this;
+        }
+
+        public RuntimeQueryResultItemBuilder with(Runtime runtime) {
+            this.runtime = runtime;
+            return this;
+        }
+
+        public RuntimeQueryResultItem build() {
+            RuntimeQueryResultItem item = new RuntimeQueryResultItem();
+
+            if (pipelineExecutorTrace != null) {
+                if (pipelineExecutorTrace.getTask().getTaskDef().getProviderId() != null) {
+                    item.setProviderId(pipelineExecutorTrace.getTask().getTaskDef().getProviderId().getId());
+                    item.setProviderTypeName(pipelineExecutorTrace.getTask().getTaskDef().getProviderId().getProviderType().getProviderTypeName());
+                    item.setProviderVersion(pipelineExecutorTrace.getTask().getTaskDef().getProviderId().getProviderType().getVersion());
+                }
+
+                item.setPipelineId(pipelineExecutorTrace.getPipelineId());
+                item.setPipelineExecutionId(pipelineExecutorTrace.getTaskId());
+                item.setPipelineStatus(pipelineExecutorTrace.getTask().getPipelineStatus().name());
+                if (pipelineExecutorTrace.getTask().getPipelineError() != null) {
+                    item.setPipelineError(pipelineExecutorTrace.getTask().getPipelineError().toString());
+                }
+
+                List<PipelineStageItem> stageItems = pipelineExecutorTrace.getTask().getTaskDef().getPipeline().getStages().stream()
+                        .map(stage -> {
+                            String stageError = null;
+                            if (pipelineExecutorTrace.getTask().getStageError(stage) != null) {
+                                stageError = pipelineExecutorTrace.getTask().getStageError(stage).getMessage();
+                            }
+                            String stageStatus = null;
+                            if (pipelineExecutorTrace.getTask().getStageStatus(stage) != null) {
+                                stageStatus = pipelineExecutorTrace.getTask().getStageStatus(stage).name();
+                            }
+                            return new PipelineStageItem(stage.getName(),
+                                                         stageStatus,
+                                                         stageError);
+                        }).collect(Collectors.toList());
+                item.setPipelineStageItems(new PipelineStageItemList(stageItems));
+
+                if (pipelineExecutorTrace.getTask().getOutput() != null &&
+                        pipelineExecutorTrace.getTask().getOutput().orElse(null) instanceof Runtime) {
+                    Runtime runtime = (Runtime) pipelineExecutorTrace.getTask().getOutput().get();
```

Comment:
Calling get on optional without checking for presence of value - might lead to Exception at runtime.

Actually there are many instances of errors like this. Could you please run Analyze > Inspect Code and see the "Probable bugs" section in IntelliJ to see how many of these are there? I'm sure some of those will lead to failures.